src/main/java/io/debezium/connector/jdbc/JdbcSinkConnectorTask.java

```diff
@@ -0,0 +1,164 @@
+ * Responsible for lifecycle management of the streaming code.
+ *
+ * @author Hossein Torabi
+ *
+ */
+public class JdbcSinkConnectorTask extends SinkTask {
+    private static enum State {
+        RUNNING,
+        STOPPED;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JdbcSinkConnectorTask.class);
+
+    private final AtomicReference<State> state = new AtomicReference<State>(State.STOPPED);
+    private final ReentrantLock stateLock = new ReentrantLock();
+    private final Map<String, Schema> topicSchema = new HashMap<>();
+
+    private JdbcSinkConnectorConfig config;
+
+    private SessionFactory sessionFactory;
+    private EntityManager entityManager;
+    private Session session;
+
+    @Override
+    public String version() {
+        return Module.version();
+    }
+
+    @Override
+    public void start(Map<String, String> props) {
+        stateLock.lock();
+
+        try {
+            if (!state.compareAndSet(State.STOPPED, State.RUNNING)) {
+                LOGGER.info("Connector has already been started");
+                return;
+            }
+
+            config = new JdbcSinkConnectorConfig(props);
+            config.validate();
+
+            buildSessionFactory();
+
+        }
+        finally {
+            stateLock.unlock();
+        }
+    }
+
+    @Override
+    public void put(Collection<SinkRecord> records) {
+        try {
+            session = entityManager.unwrap(Session.class);
+            Transaction transaction = session.beginTransaction();
+            for (SinkRecord record : records) {
+                Struct value = (Struct) record.value();
+                Schema valueSchema = record.valueSchema();
+                String topic = record.topic();
+                if (!Envelope.isEnvelopeSchema(valueSchema)) {
+                    throw new RuntimeException("Only Debezium schema is supported");
+                }
+
+                Struct afterStruct = value.getStruct(Envelope.FieldName.AFTER);
+                Schema afterSchema = afterStruct.schema();
+
+                if (!topicSchema.containsKey(topic) || !topicSchema.get(topic).equals(afterSchema)) {
+                    topicSchema.put(topic, afterSchema);
+                    JaxbHbmHibernateMapping mapping = HibernateJaxbRecordMapper.getJaxbHbmHibernateMapping(topic, afterSchema);
+                    try {
+                        persistJaxbHbmHibernateMapping(topic, mapping);
+                    }
+                    catch (JAXBException | FileNotFoundException e) {
+                        throw new RuntimeException("Can not persists new schema %s", e);
+                    }
+                    // commit old data
+                    transaction.commit();
+                    session.close();
+
+                    buildSessionFactory();
+                    session = entityManager.unwrap(Session.class);
+                    transaction = session.beginTransaction();
```

Comment:
Doesn't this multi-transactional approach become problematic if the batch of sink records gets persisted partially and then an exception is later thrown?  What happens when the sink is restarted the entire batch is re-sent?  Is the entire process going to be idempotent?

Reply:
the whole batch `Collection<SinkRecord>` in `put` method persist with only one Transaction, so there is no partially persistence. definitely it needs more effort to make it more durable and persistence.

Reply:
I don't believe that's correct.

The way the current loop is designed if you receive a record for an entity you already know about you persist it in the current transaction; however if you detect any new entity type, you commit the current transaction, build a JAXB mapping, restart the session factory and begin a new session/transaction before you persist the record.  This means that if you get a mix of records for entities you know and entities you do not know in the same batch; you have N+1 transactions that get created, where N is the number of new entity types in the sink record collection.